numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -793,6 +800,43 @@ PyArray_CanCastTypeTo(PyArray_Descr *from, PyArray_Descr *to,
                 break;
         }
     }
+    /*
+     * Casting from a structured type with fields to a simple one is only
+     * OK if there is only a single field and casting from that field works.
+     *
+     * Casting from a simple type to a structured one is only OK if casting
+     * to all fields works.
+     *
+     * Note that the case of structured to structured is already dealt with
+     * above, since those have EquivTypenum.
+     */
+    else if (PyDataType_HASFIELDS(from)) {
+        assert(!PyDataType_HASFIELDS(to));
+        if (PyDict_Size(from->fields) == 1) {
+            Py_ssize_t ppos = 0;
+            PyObject *tuple;
+            PyDict_Next(from->fields, &ppos, NULL, &tuple);
+            return PyArray_CanCastTypeTo(
+                (PyArray_Descr *)PyTuple_GET_ITEM(tuple, 0),
+                to, casting);
```

Comment:
OK, I think I disagree here, this is not safe casting, but only valid for unsafe casting.